### PR TITLE
[Query] Check whether cap is fixed or not.

### PR DESCRIPTION
**Pull request by gichan-jang**
- Check whether the cap is fixed or not before setting the cap.
 
Signed-off-by: gichan <gichan2.jang@samsung.com>

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*] Passed [ ]Failed [ ]Skipped

### CODE REVIEW

File: gst/nnstreamer/tensor_query/tensor_query_client.c [hunk 1]
```diff
@@ -342,25 +342,49 @@ gst_tensor_query_client_get_property (GObject * object, guint prop_id,
...
+    if (sink_caps) {
+      structure = gst_caps_get_structure (sink_caps, 0);
+      if (gst_structure_has_field (structure, "framerate")) {
+        gst_structure_get_fraction (structure, "framerate", &rate_n, &rate_d);
+      }
+      gst_caps_unref (sink_caps);
+    }
+
+    out_caps = gst_caps_make_writable (out_caps);
+    gst_caps_set_simple (out_caps, "framerate", GST_TYPE_FRACTION, rate_n,
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/tensor_query_client.c:371]: Q: Is it necessary to set framerate on srcpad?
src pad caps template is any. plz consider the case if src is audio or other mimetype that does not need framerate.

File: gst/nnstreamer/tensor_query/tensor_query_client.c [hunk 2]
```diff
@@ -342,25 +342,49 @@ gst_tensor_query_client_get_property (GObject * object, guint prop_id,
...
+        rate_d, NULL);
+
+    if (gst_caps_is_fixed (out_caps)) {
+      gst_pad_set_caps (self->srcpad, out_caps);
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/tensor_query_client.c:375]: ret = gst_pad_set_caps (self->srcpad, out_caps);

**Comment by gichan-jang** [comment]
> Q: Is it necessary to set framerate on srcpad?
src pad caps template is any. plz consider the case if src is audio or other mimetype that does not need framerate.

I changed to print error messages and stop the pipeline when the tensors config is invalid. (e.g., framerate is not set.)

File: gst/nnstreamer/tensor_common.c [hunk 3]
```diff
@@ -274,6 +274,8 @@ gst_tensor_info_validate (const GstTensorInfo * info)
   g_return_val_if_fail (info != NULL, FALSE);
 
   if (info->type == _NNS_END) {
+    nns_loge ("Failed to validate tensor info. type: %d. "
+        "Please specify tensor type. e.g., types=uint8,float32 ", info->type);
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_common.c:278]: Recommend to use gst_tensor_get_type_string() for log message.
```
nns_loge ("Failed to validate tensor info. type: %s. "
        "Please specify tensor type. e.g., types=uint8,float32 ", GST_STR_NULL (gst_tensor_get_type_string (info->type));
```

File: gst/nnstreamer/tensor_common.c [hunk 4]
```diff
@@ -794,11 +798,16 @@ gst_tensors_config_validate (const GstTensorsConfig * config)
...
   if (config->rate_n < 0 || config->rate_d <= 0) {
+    nns_loge ("Failed to validate tensors config. framerate: %d/%d. "
+        "framerate should be numerator >= 0 and denominator > 0.",
+        config->rate_n, config->rate_d);
     return FALSE;
   }
 
   /* tensor stream format */
   if (config->format >= _NNS_TENSOR_FORMAT_END) {
+    nns_loge ("Failed to validate tensors config. format: %d. "
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_common.c:809]: Also recommend to use gst_tensor_get_format_string() here.

File: gst/nnstreamer/tensor_query/tensor_query_serversink.c [hunk 5]
```diff
@@ -227,16 +227,17 @@ gst_tensor_query_serversink_start (GstBaseSink * bsink)
   GstTensorQueryServerSink *sink = GST_TENSOR_QUERY_SERVERSINK (bsink);
   GstCaps *caps;
   gchar *caps_str = NULL;
+  gboolean ret = FALSE;
```
Comment (dongju-chae) [inline gst/nnstreamer/tensor_query/tensor_query_serversink.c:230]: Well.. this variable seems not to be changed in runtime. Any merit to declare it?

**Comment by gichan-jang** [comment]
@jaeyun-jung @dongju-chae 
The tensors config validation check at the start of query server sink was removed. Even if the cap is not defined at start, the cap is received from the connected element during the negation phase. 
So, in this PR, I removed setting framerate at the start of the server sink and changed it to check only if it is fixed before setting the cap on the query client source pad.

>Recommend to use gst_tensor_get_type_string() for log message
```
nns_loge ("Failed to validate tensor info. type: %s. "
        "Please specify tensor type. e.g., type=uint8.",
        GST_STR_NULL (gst_tensor_get_type_string (info->type));
```

Thanks. Since the validation function is not used in this patch, the changes will be uploaded to another PR.


> Well.. this variable seems not to be changed in runtime. Any merit to declare it?
Removed related changes. Thanks.

File: gst/nnstreamer/tensor_query/tensor_query_client.c [hunk 6]
```diff
@@ -342,25 +342,33 @@ gst_tensor_query_client_get_property (GObject * object, guint prop_id,
...
   curr_caps = gst_pad_get_current_caps (self->srcpad);
   if (curr_caps == NULL || !gst_caps_is_equal (curr_caps, out_caps)) {
-    silent_debug_caps (self, out_caps, "set out-caps");
-    gst_pad_set_caps (self->srcpad, out_caps);
+    if (gst_caps_is_fixed (out_caps)) {
+      ret = gst_pad_set_caps (self->srcpad, out_caps);
+    } else {
+      nns_loge ("out-caps from tensor_query_serversink is not fixed. "
+          "Failed to update client src caps, out-caps: %s", caps_str);
+      return ret;
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/tensor_query_client.c:362]: plz remove this line so that out_caps can be released.

**Comment by gichan-jang** [comment]
> plz remove this line so that out_caps can be released.

Updated. Thanks.

**Review by dongju-chae — APPROVED** [review]
LGTM

**Review by jaeyun-jung — APPROVED** [review]

**Pull request merged**